File: res/scripts/prototype/engine.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Copyright (C) Joan Sala Soler, 2014 <contact@joansala.com>

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
"""

import sys
import threading


class Engine(object):
    """A really simple game engine"""
    
    __NAME = 'Prototype'
    __AUTHOR = 'Joan Sala Soler'
    
    __INFINITY = 1000
    __MIN_DEPTH = 4
    
    __EXACT = 0
    __LOWER = 1
    __UPPER = 2
    
    
    def __init__(self, game):
        """Initializes this engine"""
        
        self._game = game
        self._depth = 126
        self._move_time = 2.0
        self._aborted = False
        self._line = dict()
        
        
    def get_name(self):
        """Returns this engine name"""
        
        return Engine.__NAME
        
        
    def get_author(self):
        """Returns this engine author"""
        
        return Engine.__AUTHOR
        
        
    def get_depth(self):
        """Returns the maximum search depth"""
        
        return self._depth
        
        
    def get_move_time(self):
        """Returns the maximum allowed time per search"""
        
        return self._move_time
        
        
    def set_depth(self, depth):
        """Sets the maximum search depth"""
        
        self._depth = depth < Engine.__MIN_DEPTH \
            and Engine.__MIN_DEPTH \
            or int(depth) + int(depth) % 2
        
        
    def set_move_time(self, move_time):
        """Sets the maximum allowed time per search"""
        
        self._move_time = move_time
        
        
    def stop_computation(self):
        """Aborts the current computation"""
        
        self._aborted = True
        
        
    def compute_best_move(self, board, turn, positions = tuple()):
        """Computes a best move for the specified position"""
        
        # Start a move time countdown
        
        self._aborted = False
        
        t = threading.Timer(
            self._move_time,
            self.stop_computation
        )
        
        t.start()
        
        # Fill the opened line of play
        
        self._line = dict.fromkeys(positions)
        moves = tuple(self._game.xlegal_moves(board, turn))
        
        # Iterative deepening search for a best move
        
        depth = Engine.__MIN_DEPTH
        alpha = -Engine.__INFINITY
        
        while not self._aborted:
            best_move = moves[0]
            best_score = -Engine.__INFINITY
            
            for cmove in moves:
                cboard = self._game.compute_board(board, cmove)
                
                # Perform a negamax search
                
                score = -self._search(
                    cboard,
                    -turn,
                    alpha,
                    -best_score,
                    depth
                )
                
                # Make sure the minimal depth is reached
                
                if self._aborted and depth > Engine.__MIN_DEPTH:
                    if last_score >= best_score:
                        best_move = last_move
                    break
                
                # Remember best moves and score
                
                if score > best_score:
                    best_move = cmove
                    best_score = score
            
            if depth >= self._depth:
                break
            
            # She's heading for the disco…
            
            last_move = best_move
            last_score = best_score
            depth += 2
        
        t.cancel()
        
        return best_move
        
        
    def _search(self, board, turn, alpha, beta, depth):
        """Recursive search of the best moves"""
        
        if self._aborted: return -Engine.__INFINITY
        
        # If the node is terminal, return its value
        
        position = (board, turn)
        
        if self._game.is_end(board, turn) or position in self._line:
            return turn * self._game.get_final_score(board)
        
        if depth == 0:
            return turn * self._game.get_score(board)
        
        # Recursive search for a best move
        
        self._line[position] = True
        
        for cmove in self._game.xlegal_moves(board, turn):
            cboard = self._game.compute_board(board, cmove)
            
            score = -self._search(
                cboard,
                -turn,
                -beta,
                -alpha,
                depth - 1
            )
            
            if score >= beta:
                alpha = beta
                break
            
            if score > alpha:
                alpha = score
        
        del self._line[position]
        
        return alpha

File: src/main/java/com/joansala/engine/Engine.java
package com.joansala.engine;

/*
 * Copyright (C) 2014 Joan Sala Soler <contact@joansala.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import java.util.function.Consumer;


/**
 * An engine searches a game and returns a best move for its current state.
 *
 * @author    Joan Sala Soler
 * @version   1.0.0
 */
public interface Engine {

    /** Default depth limit per move */
    int DEFAULT_DEPTH = 127;

    /** Default time limit per move */
    long DEFAULT_MOVETIME = 3600;


    /**
     * Returns the maximum depth allowed for the search
     *
     * @return   The depth value
     */
    int getDepth();


    /**
     * Returns the maximum time allowed for a move computation
     * in milliseconds
     *
     * @return   The new search time in milliseconds
     */
    long getMoveTime();


    /**
     * Returns the current the contempt factor of the engine.
     */
    int getContempt();


    /**
     * Returns the current infinity score of the engine.
     */
    int getInfinity();


    /**
     * Obtains the move this engine would like to ponder on.
     */
    int getPonderMove(Game game);


    /**
     * Sets the maximum search depth for subsequent computations
     *
     * @param depth  The new depth value
     */
    void setDepth(int depth);


    /**
     * Sets the maximum search time allowed for subsequent computations
     *
     * @param delay    The new time value in milliseconds as a
     *                 positive number greater than zero
     */
    void setMoveTime(long delay);


    /**
     * Sets the contempt factor. That is, the score to which end game
     * positions that are draw will be evaluated. The contempt score
     * must be a value in the range +-infinity for the engine to work
     * properly.
     *
     * @see Engine#setInfinity(int)
     * @param score     Score for draw positions
     */
    void setContempt(int score);


    /**
     * Sets the infinity score. Setting this value to the maximum score
     * a game object can possibly be evaluated to is likely to improve
     * the engine performance.
     *
     * @param score     Infinite value as apositive integer
     */
    void setInfinity(int score);


    /**
     * Attaches a move consumer to the engine.
     *
     * Consumers may be invoked by the engine whenever some search information
     * changes. They are granteed to run till completion before the engine
     * continues with normal operations. Each consumer receives as a parameter
     * the best move found so far by the engine.
     *
     * @param consumer  Best move consumer
     */
    void attachConsumer(Consumer<Report> consumer);


    /**
     * Detaches a move consumer from the engine.
     *
     * @param consumer  Best move consumer
     */
    void detachConsumer(Consumer<Report> consumer);


    /**
     * Tells the engine that the next positions are going to be from
     * a different match. This method must be called before a new game
     * is started to ask the engine to set up its parameters acordingly.
     */
    void newMatch();


    /**
     * Best score obtainable for the current game state.
     *
     * @param game  Initial game state
     * @return      Computed score
     */
    int computeBestScore(Game game);


    /**
     * Computes a best move for the current position of a game.
     *
     * @param game  The game for which a best move must be computed
     * @return      The best move found for the current game position
     *              or {@code Game.NULL_MOVE} if the game already ended
     */
    int computeBestMove(Game game);


    /**
     * Aborts the current search. After calling this method any move
     * computations must return a result immediately.
     *
     * @see Engine#computeBestMove(Game)
     */
    void abortComputation();


    /**
     * Stop the current search after a specified delay.
     *
     * @param delay      Delay in milliseconds
     */
    void abortComputation(long delay);
}
